resolver/resolver_test.go

```diff
@@ -433,52 +433,145 @@ func TestTruncateEdns0NoSetTruncateBit(t *testing.T) {
 	}
-	// expect it to have been truncated
-	if msg2.Len() > int(maxSize) {
+	if msg.Len() > int(max) {
 		t.Fatal("message not truncated")
 	}
+	if msg.Truncated {
+		t.Fatal("Truncate bit set even though setTruncateBit was false")
+	}
+	// We set the Truncate bit on the message and confirm that the bit is
+	// cleared even though the message did not need to be truncated.
+	// This asserts that no matter the message or its size, truncate() will
+	// clear the Truncate bit if setTruncateBit=false is passed to truncate().
+	before := msg.Len()
+	msg.Truncated = true
+	truncate(msg, max, false)
+	if msg.Truncated {
+		t.Fatal("Truncate bit not cleared")
+	}
+	if msg.Len() != before {
+		t.Fatal("message truncated further")
+	}
+	msg.Answer = append(msg.Answer, genA(1)...)
+	if l := msg.Len(); l < int(max) {
+		t.Fatalf("Message to small after adding answers: %d bytes", l)
+	}
+}
+
+func TestTruncateAnswers(t *testing.T) {
+	// Test truncating messages by starting with empty answers and increasing
+	// the number of answers until twice the maximum size is reached and test
+	// that certain invariants are maintained at all times.
+	// Testing boundary conditions only would be typical if the problem space
+	// was large, however in this case since the maximum size is so small
+	// we test the space exhaustively.
+	max := uint16(4096)
+	for nn := 0; ; nn++ {
+		msg := Message(
+			Question("example.com.", dns.TypeA),
+			Header(false, dns.RcodeSuccess),
+			Answers(genA(nn)...))
+		before := msg.Len()
+		if before > int(max)*2 {
+			// We've really exhausted the problem space by creating
+			// messages that have minimal size, all the way too messages
+			// that are way bigger they are allowed to be.
+			return
+		}
+		if before <= int(max) {
+			// This message is not too large.
+			// Assert that we don't truncate messages that
+			// don't need to be truncated
+			truncateAnswers(msg, max)
+			if msg.Len() != before {
+				t.Fatal("small message truncated further")
+			}
+			// Add another answer to check for the case where adding another
+			// answer pushes the message size over the max limit. We expect
+			// truncateAnswers to drop the last answer.
+			msg.Answer = append(msg.Answer, genA(1)...)
+			if msg.Len() > int(max) {
+				truncateAnswers(msg, max)
+				if msg.Len() != before {
+					t.Fatal("truncateAnswers did not drop the last answer.")
+				}
+			}
+			continue
+		}
+		// This message is too large, we truncate its list of answers and
+		// check that it narrowly fits afterwards.
+		truncateAnswers(msg, max)
+		if msg.Len() > int(max) {
+			t.Fatal("truncateAnswers did not truncate enough")
+		}
+		msg.Answer = append(msg.Answer, genA(1)...)
+		if msg.Len() <= int(max) {
+			t.Fatal("truncateAnswers truncated the message too much")
+		}
+	}
+}
+
+func TestMaxMsgSizeTCP(t *testing.T) {
+	msg := newTruncated()
+	msg.SetEdns0(4096, false)
+	edns0 := msg.IsEdns0()
+	if edns0 == nil {
+		panic("SetEdns0 did not set the EDNS0 (OPT) record")
```

Comment:
Yes - I spent some time deciding so here are my thoughts.

If `edns0` is nil, it means the `dns` library is broken. It does not mean the behaviour that is being tested has failed. It is closer to a runtime error than a failed assert.